### PR TITLE
perf: compress and hardlink trampolines

### PR DESCRIPTION
**Pull request by nichmor**
This PR introduce two small optimisations:

* first we store compressed trampolines by simple using zstd. 
This give us a room to save almost 50% of size for each of the trampoline when including them in pixi
`trampolines/pixi-trampoline-aarch64-apple-darwin : 50.61%   (   377 KiB =>    191 KiB, trampolines/pixi-trampoline-aarch64-apple-darwin.zst)`

* instead of copying it every time, we will write it only once and then just hardlink to original trampoline.
original trampoline will be saved under 
`/Users/graf/.pixi/bin/trampoline_configuration/trampoline_bin`.

global binaries when installed will point to this inode as we can see in screenshot:
Screenshot 2024-11-05 at 11 35 07

Screenshot 2024-11-05 at 11 35 12

p.s.

I've got inspiration for the tests for hardlink from here:
https://github.com/rust-lang/rust/blob/27e38f8fc7efc57b75e9a763d7a0ee44822cd5f7/library/std/src/fs/tests.rs#L949

### CODE REVIEW

File: .github/workflows/trampoline.yaml [hunk 1]
```diff
@@ -85,6 +85,12 @@ jobs:
           mkdir -p trampolines
           # Iterate through all files in trampolines directory and its subdirectories
           find trampolines-binaries -type f  -name 'pixi-trampoline-*' -exec mv -f {} trampolines/ \;
+          # now iterate through all files in trampolines directory and compress them using zstd
+          # and remove the original file
+          for file in trampolines/*; do
```
[the remark this answers is not in the document]
Comment (Hofer-Julian) [reply .github/workflows/trampoline.yaml:90]: Indeed, seems like `zstd` doesn't provide a macro for that.

**Review by ruben-arts — COMMENTED** [review]
Look great, could you document this behavior so people can search on `trampoline` in the documentation and get this logic explained.

**Comment by nichmor** [comment]
> Look great, could you document this behavior so people can search on `trampoline` in the documentation and get this logic explained.

added module doc

File: crates/pixi_trampoline/README.md [hunk 2]
```diff
@@ -18,5 +18,6 @@ This is the configuration used by trampoline to set the env variables, and run t
 
 # How to build it?
 You can use `trampoline.yaml` workflow to build the binary for all the platforms and architectures supported by pixi.
-If running it manually from forked repo, you can download and commit artifacts to your branch.
+In case of building it manually, you can use the following command, after executing the `cargo build --release`, you need to compress it using `zstd`.
+Then you can commit `compressed` artifacts to your branch.
```
Comment (Hofer-Julian) [inline crates/pixi_trampoline/README.md:22]:
```suggestion
```

We don't really want people to commit it manually, right?
Maybe let's skip that sentence?

**Review by Hofer-Julian — APPROVED** [review]
Looks great and works as expected. Great job!

**Pull request merged**